contracts/coordinator/src/contract/execute.rs

```diff
@@ -55,3 +78,243 @@ pub fn set_active_verifier_set(
         .change_context(Error::VerifierSetActivationFailed)?;
     Ok(Response::new())
 }
+
+fn instantiate2_addr(
+    deps: &DepsMut,
+    env: &Env,
+    code_id: u64,
+    salt: &[u8],
+) -> error_stack::Result<Addr, Error> {
+    let code_info: cosmwasm_std::CodeInfoResponse = deps
+        .querier
+        .query(&WasmQuery::CodeInfo { code_id }.into())
+        .change_context(Error::QueryCodeInfo(code_id))?;
+
+    deps.api
+        .addr_humanize(
+            &cosmwasm_std::instantiate2_address(
+                code_info.checksum.as_slice(),
+                &deps
+                    .api
+                    .addr_canonicalize(&env.contract.address.to_string().clone())
+                    .change_context(Error::Instantiate2Address)?,
+                salt,
+            )
+            .change_context(Error::Instantiate2Address)?,
+        )
+        .change_context(Error::Instantiate2Address)
+}
+
+fn launch_contract(
+    deps: &DepsMut,
+    info: &MessageInfo,
+    env: &Env,
+    salt: Binary,
+    code_id: u64,
+    instantiate_msg: Binary,
+    label: String,
+) -> error_stack::Result<(WasmMsg, Addr), Error> {
+    Ok((
+        WasmMsg::Instantiate2 {
+            admin: Some(info.sender.to_string()),
+            code_id,
+            msg: instantiate_msg,
+            funds: info.funds.clone(),
+            label,
+            salt: salt.clone(),
+        },
+        instantiate2_addr(deps, env, code_id, salt.as_slice())?,
+    ))
+}
+
+fn instantiate_gateway(
+    ctx: &InstantiateContext,
+    label: String,
+    router_address: Addr,
+    verifier_address: Addr,
+) -> error_stack::Result<(WasmMsg, Addr), Error> {
+    launch_contract(
+        &ctx.deps,
+        &ctx.info,
+        &ctx.env,
+        ctx.salt.clone(),
+        ctx.gateway_code_id,
+        cosmwasm_std::to_json_binary(&gateway_api::msg::InstantiateMsg {
+            verifier_address: verifier_address.to_string().clone(),
+            router_address: router_address.to_string().clone(),
+        })
+        .change_context(Error::InstantiateGateway)?,
+        label,
+    )
+}
```

Comment:
I think these instantiation functions and the way they are called could be improved, but I need to think a bit more about it